inst/training_202307/202307.qmd

```diff
@@ -0,0 +1,448 @@
++ refs:
+  - dplyr@1.0.0
+  - tibble@1.0.0
++ constraints (93):
+  (...)
+x failures:
+* dplyr@1.0.0: Can't install dependency tibble (>= 2.1.3)
+* tibble@1.0.0: Conflicts with tibble@1.0.0
+* tibble: Conflicts with tibble@1.0.0
+```
+
+## Example - success
+
+:::: {.columns}
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="8"}
+Type: Package
+Package: foo.package
+Version: 1.2.3.9000
+Depends:
+    R (>= 3.6)
+Imports:
+    dplyr (>= 1.0.0),
+    tibble (>= 2.1.3)
+    (...)
+```
+:::
+
+::: {.column width="50%"}
+```{.yaml code-line-numbers="8"}
+Type: Package
+Package: dplyr
+Package: 1.0.0
+Depends:
+    R (>= 3.2.0)
+Imports:
+    (...)
+    tibble (>= 2.1.3)
+    (...)
+```
+:::
+
+::::
+
+<br>
+
+```{r}
+#| eval: false
+#| echo: true
+#| code-line-numbers: false
+x <- pkgdepends::new_pkg_deps(c("dplyr@1.0.0", "tibble@2.1.3"))
+x$solve()
+x$get_solution()
+```
+
+```.default
+<pkg_solution>
++ result: OK
++ refs:
+  - dplyr@1.0.0
+  - tibble@2.1.3
++ constraints (95):
+  - select dplyr exactly once
+  - select tibble exactly once
+  - select R6 at most once
+  - select cli at most once
+  - select ellipsis at most once
+  - select fansi at most once
+  - select generics at most once
+  - select glue at most once
+  - select lifecycle at most once
+  - select magrittr at most once
+  ...
++ solution:
+  - cli
+  - crayon
+  - dplyr@1.0.0
+  - ellipsis
+  - fansi
+  - generics
+  - glue
+  - lifecycle
+  - magrittr
+  - pillar
+  - pkgconfig
+  - R6
+  - rlang
+  - tibble@2.1.3
+  - tidyselect
+  - utf8
+  - vctrs
+  - withr
+```
+
+## Discussion:
+
+- One might argue that `foo.package` is fully compatible with the `tibble@1.0.0` syntax but the truth is that `tibble@1.0.0` cannot be installed alongside `dplyr@1.0.0` hence `foo.package` cannot be used with both `dplyr@1.0.0` and `tibble@1.0.0` loaded. Therefore `tibble (>= 1.0.0)` really indicates `tibble (>= 2.1.3)` in this case.
+
+- Because of above, the minimal version specification of dependencies is the _true_ **installable** and **compatible** set of dependencies.
+
+## Real-life observations:
+
+- Oftentimes the dependency tree is of a big depth and there are multi-step conflicts.
+
+- Some very old packages are not compilable on modern machines and you have to increase versions to make it installable.
+
+- The package resolution is just the first step. The next one is to use them in `R CMD CHECK` and validate its compatibility.
+
+- In order to enter GHA debug lines, `verdepcheck` will continue on error entering `R CMD CHECK` step when e.g. dependency resolve failed. If you encounter `vignette builder 'knitr' not found` error - most likely you failed on resolve or install.
+
+# A few caveats:
+
+None of the test result is _stable_:
+
+  - **min**
+
+    Indirect dependencies are installed as usual using the newest available. Therefore: it is not guaranteed that the new release of indirect dependency will be fully compatible with some old version of direct dependency.
```

Comment:
This honestly is a pretty fatal flaw in the paradigm, as I mentioned above, but it is one that you can get around with some work that is already done in switchr.